### ISSUE TITLE
<limit_maxagents> has no effect

### ISSUE THREAD

**Issue by jctello** [wazuh/wazuh#4199]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 3.10.2 | ossec-authd | Manager | Packages | OS version |

With `<limit_maxagents>` enabled (as it is by default) the manager will still register more than the amount of agents it can handle (limit mentioned here: https://github.com/wazuh/wazuh/issues/3942 ) 

To reproduce this simply run: 
`for j in {1..13998}; do /var/ossec/bin/agent-auth -m [ManagerAddress] -A 'agent'$j; done` 
From an agent to a manager and after it is done (a few minutes) see in the manager: 
```
#tail -4 /var/ossec/logs/ossec.log
2019/11/06 17:45:15 ossec-remoted: INFO: (1409): Authentication file changed. Updating.
2019/11/06 17:45:16 ossec-remoted: INFO: (1410): Reading authentication keys file.
2019/11/06 17:45:16 ossec-remoted: ERROR: (4110): Maximum number of agents '13998' reached.
2019/11/06 17:45:16 ossec-remoted: CRITICAL: (1202): Configuration error at '/etc/client.keys'.
# wc /var/ossec/etc/client.keys 
  13998   55992 1185625 /var/ossec/etc/client.keys
```

It may be an off by one error as registering agent 13999 one gets: 
```
2019/11/06 17:58:57 agent-auth: ERROR: The maximum number of agents has been reached (from manager)
2019/11/06 17:58:57 agent-auth: ERROR: Unable to add agent. (from manager)
2019/11/06 17:58:57 agent-auth: ERROR: Unable to create key. Either wrong password or connection not accepted by the manager
```
and on the manager's ossec.log:
```
2019/11/06 17:58:57 ossec-authd: INFO: New connection from 192.168.77.20
2019/11/06 17:58:57 ossec-authd: INFO: Received request for a new agent (centOS-agent16025) from: 192.168.77.20
2019/11/06 17:58:57 ossec-authd: ERROR: (4110): Maximum number of agents '13998' reached.
```

**Comment by jctello** [comment]
This issue is still present in 3.11.2 (just at a higher agent number):
```
2020/02/04 19:07:44 ossec-authd: INFO: New connection from 192.168.78.21
2020/02/04 19:07:44 ossec-authd: INFO: Received request for a new agent (agent100000) from: 192.168.78.21
2020/02/04 19:07:45 ossec-authd: INFO: Agent key generated for 'agent100000' (requested by 192.168.78.21)
2020/02/04 19:07:45 ossec-remoted: INFO: (1409): Authentication file changed. Updating.
2020/02/04 19:07:48 wazuh-modulesd: ERROR: (4110): Maximum number of agents '99998' reached.
2020/02/04 19:07:48 wazuh-modulesd: CRITICAL: (1202): Configuration error at '/var/ossec/etc/client.keys'.
2020/02/04 19:07:48 wazuh-modulesd:syscollector: INFO: Module finished.
2020/02/04 19:10:08 ossec-remoted: INFO: (1410): Reading authentication keys file.
2020/02/04 19:10:11 ossec-remoted: ERROR: (4110): Maximum number of agents '99998' reached.
2020/02/04 19:10:11 ossec-remoted: CRITICAL: (1202): Configuration error at '/etc/client.keys'.
2020/02/04 19:12:52 ossec-authd: INFO: New connection from 192.168.78.21
2020/02/04 19:12:52 ossec-authd: INFO: Received request for a new agent (agent100001) from: 192.168.78.21
2020/02/04 19:12:52 ossec-authd: ERROR: (4110): Maximum number of agents '99998' reached.
```
Note how remoted and modulesd fail before authd has output the error of too many agents.

I seems like an off by one error that would be fixed by changing the value from 2 to 3 in this logic:
https://github.com/wazuh/wazuh/blob/3.11/src/os_auth/main-server.c#L1033-L1035

This is due to the fact that the same logic is used by the modules to check for errors here:
https://github.com/wazuh/wazuh/blob/3.11/src/os_crypto/shared/keys.c#L298-L300
But the agent uses the same criteria *before* adding the agent.

**Comment by vikman90** [comment]
@jctello Thanks for fixing this issue.

This fix does not apply to local requests (`manage_agents` or API). I've ported it to the common Authd library so that it applies to both.

However, IMO the bug is not in Authd (for accepting `MAX_AGENTS - 2` agents) but in Remoted and Modulesd (for rejecting `MAX_AGENTS - 2` agents). I've fixed the limit checking in the common agent key management library.

- [X] Authd / Valgrind: successfully load `MAX_AGENTS - 2` keys and reject further requests.
- [X] Modulesd / Valgrind: successfully sync up tp `MAX_AGENTS - 2` agent keys.
- [X] Remoted / Valgrind: successfully load `MAX_AGENTS - 2` keys.
- [X] Check unit tests.
- [X] Update changelog.

Cheers!
